src/core/io_queue.cc

```diff
@@ -365,34 +326,79 @@ class queued_read_write_request : public queued_io_request {
         : queued_io_request(q, q.request_stream(dnl), cap)
         , _sink(sink)
         , _io_request(std::move(req))
-        , _desc(std::make_unique<io_desc_read_write>(_ioq, pc, _stream, dnl, cap, std::move(iovs)))
-    {}
-
-    ~queued_read_write_request() override = default;
+        , _pclass(pc)
+        , _ts(io_queue::clock_type::now())
+        , _dnl(dnl)
+        , _iovs(std::move(iovs))
+        , _cancelled(false)
+    {
+        io_log.trace("dev {} : req {} queue  len {} capacity {}", _ioq.dev_id(), fmt::ptr(this), _dnl.length(), _fq_entry.capacity());
+    }
 
-    queued_read_write_request(queued_io_request&&) = delete;
+    // Destroys the request.
+    ~queued_read_write_request() = default;
 
+    // The queued request is neither copyable nor movable.
+    queued_read_write_request(const queued_read_write_request& other) = delete;
+    queued_read_write_request(queued_read_write_request&& other) = delete;
+    queued_read_write_request& operator=(const queued_read_write_request& other) = delete;
+    queued_read_write_request& operator=(queued_read_write_request&& other) = delete;
+
+    // Implements the interface required by queued_io_request base class.
+    // If the request was not cancelled, then submits it to io_sink.
+    // Else, finalizes the execution via io_queue and destroys 'this' object.
     void dispatch() noexcept override {
-        if (is_cancelled()) {
+        if (_cancelled) {
             _ioq.complete_cancelled_request(*this);
             delete this;
             return;
         }
 
         _intent.maybe_dequeue();
-        _desc->dispatch();
+
+        io_log.trace("dev {} : req {} submit", _ioq.dev_id(), fmt::ptr(this));
+        auto now = io_queue::clock_type::now();
+        _pclass.on_dispatch(_dnl, std::chrono::duration_cast<std::chrono::duration<double>>(now - _ts));
+        _ts = now;
+
         _ioq.request_dispatched();
-        _sink.submit(_desc.release(), std::move(_io_request));
-        delete this;
+        _sink.submit(this, std::move(_io_request));
```

Comment:
The _io_request part is not duplicated inside sink. First, it lives on read-write request here (in moved-from form, but it still occupies memory), second as a part of pending_io_request

Reply:
Could you elaborate a little bit?

Previously `queued_io_request` was derived from `internal::io_request`. Private inheritance was used -- it models `has-a` relationship that is almost like having a member variable.

Because we neither expose any members of this base class via using directive nor implement empty-base optimization idiom I concluded that having private inheritance in code is a little bit distracting for a reader -- using member variable was much simpler to understand.

The statement about memory usage is true. Previously, `queued_io_request` was destroyed after it was dispatched. Thus, its `internal::io_request` sub-object was removed. Currently, it is not like that - we do not remove the object right after the dispatch (if it was submitted to io_sink).

Do you propose to allocate `_io_request` on heap? This way the object would be freed after the dispatch. In such case `_cancelled` variable could be removed -- we could use the pointer to `_io_request` to check whether the request can be dispatched.


Also one note about moved-from state of `internal::io_request` -- this type is a tagged union that contains only POD structs -- thus as far as I understand move is just a copy -- we cannot "steal" any resources from the original object.

Reply:
There are currently 4 objects involved in IO:
- internal::io_request -- it's an API thing that users of IO queue use to tell what IO they want to do. BTW, I think you need to extend it and add operation::discard into it
- queued_io_request -- it's used to track a request that sits in seastar queue. It has fair_queue entry to track its queueing, it has internal::io_request itself, because this is what this request is doing and it has io_desc_read_write (see below)
- pending_io_request -- it's used to track a request that sits in the sink, i.e. -- ready to be sent to the kernel. It has internal::io_request itfels, because this is what this request is doing, and (via io_completion) the io_desc_read_write (see below)
- io_desc_read_write -- it's a completion tracker that's implicitly attached to internal::io_request once it gets into io-queue and that's used to inform all the interested parties after the request completes

My point is that queued_io_request and pending_io_request has their well defined role and life-time -- they help internal::io_request and io_completion (backed by io_desc_read_write) travel through io stack -- the former is there while the request is queued the latter is there while the request is being dispatched (very short). Also queued_io_request is cancellable, while pending_io_request isn't.

This change breaks this harmony by making queued_io_request combine several roles and extending its lifetime. It's not bad on its own, my point is mostly about pending_io_request -- after the above change it's no longer clear what it's needed for.

Reply:
> Also one note about moved-from state of internal::io_request -- this type is a tagged union that contains only POD structs -- thus as far as I understand move is just a copy -- we cannot "steal" any resources from the original object.

Yes, of course, my concern was that "it still occupies memory"

Reply:
Okay, thank you for the feedback :+1: 

>internal::io_request -- it's an API thing that users of IO queue use to tell what IO they want to do. BTW, I think you need to extend it and add operation::discard into it

Sure, I have just prepared a PR with the suggested change -- please find https://github.com/scylladb/seastar/pull/2330

Regarding the adjustments of `io_queue` -- I will rework the code according to your comments to meet the required criteria. Also, locally I tried to move forward with `queued_discard_request` implementation. I ended up with a class, that it is very similar to `queued_read_write_request` -- majority of the functionality was identical (except the fact that it was called from different places).

I will rework this PR.